database/interface.py

```diff
@@ -142,13 +149,117 @@ def update_harvest_source(self, source_id, updates):
+            return "Harvest source has no records to clear."
+
+        ckan_ids = [record.ckan_id for record in records if record.ckan_id is not None]
+        error_records = [record for record in records if record.status == 'error']
+        jobs_in_progress = self.get_all_harvest_jobs_by_filter(
+            {"harvest_source_id": source.id, "status": "in_progress"}
+        )
+
+        # Ensure no jobs are in progress
+        if jobs_in_progress:
+            return ("Error: A harvest job is currently in progress. "
+                    "Cannot clear datasets.")
+
+        # Ensure (error_records + ckan_ids) = total records
+        if len(error_records) + len(ckan_ids) != len(records):
+            return ("Error: Not all records are either in an error state "
+                    "or have a CKAN ID. Cannot proceed without clearing the dataset.")
+
+        if not ckan_ids:  
+            _clear_harvest_records()
+            return "Harvest source cleared successfully."
+
+        ckan = RemoteCKAN(
+            os.getenv("CKAN_API_URL"), apikey=os.getenv("CKAN_API_TOKEN")
+        )
+
+        result = ckan.action.package_search(fq=f'owner_org:{organization_id}')
+        ckan_datasets = result['count']
+        start = datetime.now(timezone.utc)
+        retry_count = 0
+        retry_max = 20
+
+        # Retry loop to handle timeouts from cloud.gov and CKAN's Solr backend, 
+        # ensuring datasets are cleared despite possible interruptions.
+        while ckan_datasets > 0 and retry_count < retry_max:
+            result = ckan.action.package_search(fq=f'owner_org:{organization_id}')
+            ckan_datasets = result['count']
+            logger.info(f"Attempt {retry_count + 1}: "
+                        f"{ckan_datasets} datasets remaining in CKAN")
+            try:
+                ckan.action.bulk_update_delete(
+                    datasets=ckan_ids, org_id=organization_id
+                )
+            except ckanapi.errors.CKANAPIError as api_err:
+                logger.error(f"CKAN API error: {api_err}")
+            except Exception as err:
+                logger.error(f"Error occurred: {err} \n error_type: {type(err)}")
+                return f"Error occurred: {err}"
+
+            retry_count += 1
+            time.sleep(5)
+
+        # If all datasets are deleted from CKAN, clear harvest records
+        if ckan_datasets == 0:
+            logger.info("All datasets cleared from CKAN, clearing harvest records.")
+            _clear_harvest_records()
+            logger.info(f"Total time: {datetime.now(timezone.utc) - start}")
+            return "Harvest source cleared successfully."
+        else:
+            fail_message = (f"Harvest source clearance failed after {retry_count} "
+                            f"attempts. {ckan_datasets} datasets still exist in CKAN.")
+            logger.error(fail_message)
+            return fail_message
+
     def delete_harvest_source(self, source_id):
```

Comment:
this is only for records on the flask side?
shouldn't we clear the CKAN source before deleting?